packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -0,0 +1,204 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import type { Type } from 'typescript';
+
+import * as util from '../util';
+
+export type Options = [
+  {
+    ignoreIntersections?: boolean;
+    ignoreUnions?: boolean;
+  },
+];
+
+export type MessageIds = 'duplicate';
+
+const isRecordType = (
+  object: object,
+): object is Record<string | symbol, unknown> => {
+  return object.constructor === Object;
+};
+
+const astIgnoreKeys = ['range', 'loc', 'parent'];
+
+const isSameAstNode = (actualNode: unknown, expectedNode: unknown): boolean => {
+  if (actualNode === expectedNode) {
+    return true;
+  }
+  if (
+    actualNode &&
+    expectedNode &&
+    typeof actualNode == 'object' &&
+    typeof expectedNode == 'object'
+  ) {
+    if (actualNode.constructor !== expectedNode.constructor) {
+      return false;
+    }
```

Comment:
Indeed, there are a few lines of code not covered by tests. Which means there should be tests covering those lines, or if that's not possible, those lines can be removed.

My other comments in this review have some test suggestions in them. I'm also thinking maybe:
* `number[]` vs `string[]`
* `string[]` vs `string[]`
* `number[][]` vs `string[]`
* `string[][]` vs `string[][]`
* `ClassA<number>` vs `ClassB<number>`
* `ClassA<number>` vs `ClassB<string>`

Are there any others you're thinking of?